clients/src/main/java/org/apache/kafka/clients/admin/internals/FenceProducersHandler.java

```diff
@@ -0,0 +1,146 @@
+    private final Logger log;
+    private final AdminApiLookupStrategy<CoordinatorKey> lookupStrategy;
+
+    public FenceProducersHandler(
+            LogContext logContext
+    ) {
+        this.log = logContext.logger(DescribeTransactionsHandler.class);
+        this.lookupStrategy = new CoordinatorStrategy(FindCoordinatorRequest.CoordinatorType.TRANSACTION, logContext);
+    }
+
+    public static AdminApiFuture.SimpleAdminApiFuture<CoordinatorKey, ProducerIdAndEpoch> newFuture(
+            Collection<String> transactionalIds
+    ) {
+        return AdminApiFuture.forKeys(buildKeySet(transactionalIds));
+    }
+
+    private static Set<CoordinatorKey> buildKeySet(Collection<String> transactionalIds) {
+        return transactionalIds.stream()
+                .map(CoordinatorKey::byTransactionalId)
+                .collect(Collectors.toSet());
+    }
+
+    @Override
+    public String apiName() {
+        return "fenceProducer";
+    }
+
+    @Override
+    public AdminApiLookupStrategy<CoordinatorKey> lookupStrategy() {
+        return lookupStrategy;
+    }
+
+    @Override
+    InitProducerIdRequest.Builder buildSingleRequest(int brokerId, CoordinatorKey key) {
+        if (key.type != FindCoordinatorRequest.CoordinatorType.TRANSACTION) {
+            throw new IllegalArgumentException("Invalid group coordinator key " + key +
+                    " when building `InitProducerId` request");
+        }
+        InitProducerIdRequestData data = new InitProducerIdRequestData()
+                .setProducerEpoch(ProducerIdAndEpoch.NONE.epoch)
+                .setProducerId(ProducerIdAndEpoch.NONE.producerId)
+                .setTransactionalId(key.idValue)
+                // Set transaction timeout to 1 since it's only being initialized to fence out older producers with the same transactional ID,
+                // and shouldn't be used for any actual record writes
+                .setTransactionTimeoutMs(1);
+        return new InitProducerIdRequest.Builder(data);
+    }
+
+    @Override
+    public ApiResult<CoordinatorKey, ProducerIdAndEpoch> handleSingleResponse(
+            Node broker,
+            CoordinatorKey key,
+            AbstractResponse abstractResponse
+    ) {
+        InitProducerIdResponse response = (InitProducerIdResponse) abstractResponse;
+
+        Errors error = Errors.forCode(response.data().errorCode());
+        if (error != Errors.NONE) {
+            return handleError(key, error);
+        }
+
+        Map<CoordinatorKey, ProducerIdAndEpoch> completed = Collections.singletonMap(key, new ProducerIdAndEpoch(
+                response.data().producerId(),
+                response.data().producerEpoch()
+        ));
+
+        return new ApiResult<>(completed, Collections.emptyMap(), Collections.emptyList());
+    }
+
+    private ApiResult<CoordinatorKey, ProducerIdAndEpoch> handleError(CoordinatorKey transactionalIdKey, Errors error) {
+        switch (error) {
```

Comment:
There's also an expected error: `PRODUCER_FENCED` which should be handled here.

Reply:
This shouldn't happen if we don't include a producer epoch/ID in the `InitProducerIdRequest`, right? If so, I could see there still being a case for adding coverage for `PRODUCER_FENCED` just in order to provide a better user-facing error message, but it's unclear what else we could really say beyond "unexpected error."

Thoughts?

Reply:
I agree with you that since we always sent without producerID and epoch, the error should never happen. But I think there might be someone to extend the existing `FenceProducerHandler` for other purpose, the assumption will not be wrong. Or at least, we can add a comment here to mention we skip `PRODUCER_FENCED` on purpose.

Reply:
Ah, good point about extensibility. I'd personally opt to just leave a comment (and in fact, remove the existing cases for `TRANSACTIONAL_ID_NOT_FOUND` and `INVALID_PRODUCER_EPOCH`, which should also never happen) in order to highlight the fact that, should those errors surface, they are completely unexpected.

If we end up needing to leverage this class somewhere else in the future, those assumptions should still hold unless we also change the `InitProducerIdRequest` we send to the broker. I can make note of that in another comment in the `buildSingleRequest` method so that we know to update the error handling logic in that case.